### ISSUE TITLE
nvim process doesn't get killed

### ISSUE THREAD

**Issue by jameh** [glacambre/firenvim#493]
I am running the latest master commit 38d8d78 on both my nvim plugin and browser extension installed manually. I am running chromium 80.0.3987.149 on Linux

When I launch chromium via keyboard shortcut (sxhkd run as a systemd service), a headless nvim process is spawned:

```
/usr/bin/nvim --headless --cmd let g:started_by_firenvim = v:true -c call firenvim#run()
```

When I tell sxhkd to stop or restart (and kill its spawned processes), it successfully kills all processes except this one. I have to manually kill the process via `pkill nvim` to continue with managing my sxhkd service.

Closing chromium also does not cause this process to die.

**Comment by glacambre** [comment]
I can reproduce this, but only with Chromium. I guess Firefox cleans up after itself when closing. Thanks for reporting this issue :)

**Comment by glacambre** [comment]
Could you try #494 and let me know if it fixes your problem? Looks like it does on my machine :)

**Comment by jameh** [comment]
Just tried it, doesn't seem to fix the problem on my computer :/

**Comment by glacambre** [comment]
Could you append `2>>/tmp/firenvim_errors` to the `exec` line of the script in `~/.local/share/firenvim/firenvim`, reproduce the bug and save `/tmp/firenvim_errors` somewhere before killing the lingering neovim process?

**Comment by jameh** [comment]
no stderr unless I give the `-V` option to nvim, then we get this output:

```
chdir(/run/user/1000/firenvim)
chdir(/etc/xdg//nvim)
chdir(/run/user/1000/firenvim)
sourcing "/etc/xdg//nvim/sysinit.vim"
chdir(/run/user/1000/firenvim)
chdir(/usr/share/nvim)
chdir(/run/user/1000/firenvim)
line 2: sourcing "/usr/share/nvim/archlinux.vim"
finished sourcing /usr/share/nvim/archlinux.vim
continuing in /etc/xdg/nvim/sysinit.vim
finished sourcing /etc/xdg//nvim/sysinit.vim
chdir(/run/user/1000/firenvim)
chdir(/home/jamie/.config/nvim)
chdir(/run/user/1000/firenvim)
sourcing "/home/jamie/.config/nvim/init.vim"
Searching for "autoload/plug.vim" in "/home/jamie/.config/nvim,/etc/xdg/nvim,/home/jamie/.local/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/usr/share/nvim/runtime,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/home/jamie/.local/share/nvim/site/after,/etc/xdg/nvim/after,/home/jamie/.config/nvim/after,/usr/share/vim/vimfiles"
Searching for "/home/jamie/.config/nvim/autoload/plug.vim"
chdir(/run/user/1000/firenvim)
chdir(/home/jamie/.config/nvim/autoload)
```

**Comment by glacambre** [comment]
What neovim version are you using and how did you get it? (compiled it yourself, installed with your package manager, downloaded from the release pages…?)

**Comment by glacambre** [comment]
I made another attempt in #494 - please try it out and let me know if I got it right this time :)

**Comment by jameh** [comment]
```
nvim -v
NVIM v0.4.3
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.4.3/src -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

I tried the new commit on branch `fix_493`, but unfortunately the process still hangs around after closing chromium!

I am being sure to build the extension every time using

```
npm install
npm run build
npm run install_manifests
```

and then I add stderr reporting to the script at `~/.local/share/firenvim`, but like before it doesn't report errors

**Comment by glacambre** [comment]
>I am being sure to build the extension every time using

Do you also update the neovim plugin?

**Comment by jameh** [comment]
are you not able to reproduce the issue with your new commits?

**Comment by jameh** [comment]
yes - I have the extension build in a post-install hook with vim-plug:

```
Plug 'glacambre/firenvim', {
  \ 'branch': 'fix_493',
  \ 'do': 'npm install && npm run build && npm run install_manifests' }
```

**Comment by jameh** [comment]
I'm using the following commands to diagnose the problems:

```
pgrep -f "nvim --headless" -a
pkill -TERM -f "nvim --headless"
```

**Comment by jameh** [comment]
So it seems on the current revision #494, the following I observe:

* kill all `nvim --headless` processes:
```
pkill -KILL -f "nvim --headless
```
* start chromium
* check that the process is spawned
```
pgrep -f "nvim --headless" -a
19397 /usr/bin/nvim --headless --cmd let g:started_by_firenvim = v:true -c call firenvim#run()
```
* close chromium
* check for the process
```
pgrep -f "nvim --headless" -a
```
* it's gone!
* redo the above procedure
* it persists!
* try to terminate it nicely
```
pkill -TERM -f "nvim --headless"
```
* check if it's still there
```
pgrep -f "nvim --headless" -a
19397 /usr/bin/nvim --headless --cmd let g:started_by_firenvim = v:true -c call firenvim#run()
```
* it's still there, wait 20s, it still hasn't shut down itself
* kill it not nicely:
```
pkill -KILL -f "nvim --headless"
```

**Comment by glacambre** [comment]
Just realized that you're probably hitting https://github.com/neovim/neovim/issues/11820 - could you try neovim nightly?

**Comment by jameh** [comment]
so, I installed neovim nightly:
```
nvim --version
NVIM v0.5.0-418-g87d892afa
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-5 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/travis/build/neovim/bot-ci/build/neovim/build/config -I/home/travis/build/neovim/bot-ci/build/neovim/src -I/home/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/home/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/home/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@travis-job-c48d5551-5074-4a05-b117-d6e98b5fddaa

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"

Run :checkhealth for more info
```

with `master` branch and with `fix_493`, the process still seems to hang around for me

**Comment by glacambre** [comment]
I believe I managed to reproduce your issue on my machine. I attached GDB to the hanged process and it looks like there's an error when trying to leave neovim - neovim then helpfully emits an error message on stderr and waits for you to press "<CR>". Of course you can't press `<CR>` since stdin and stderr have been closed.

This is caused by https://github.com/neovim/neovim/issues/9358 . I'll try to fix this and open a PR in Neovim, but it might take a while before it gets merged.

**Comment by glacambre** [comment]
I merged #494 because it can't hurt and fixing the bug in neovim is probably going to take me a long time - I managed to make a patch that solves this specific issue but it completely breaks neovim's testsuite.

**Comment by glacambre** [comment]
I merged another commit that might or might not help with this issue - please update the neovim plugin to master and let me know if it works (the browser addon will probably complain about the browser addon and the neovim plugin not having the same version but this shouldn't be an issue in theory).

**Comment by jameh** [comment]
using the latest master, the process gets killed reliably as far as I can tell!

What I did: 

* `killall nvim`, `pgrep -fa nvim` - no processes
* open chromium, `pgrep -fa nvim` - there is always at least one, more as I open edit windows
* close chromium, `pgrep -fa nvim` - no processes.

because I was curious, I tested a couple previous revisions and it appears that it's commit a80265133a351e9a6aeeb517c0f08c94879d9493 which fixes the problem - which I'm not sure if that's the exact revision I tested above and said it sometimes persisted, but it seems to reliably get killed with that revision for me now.

I'm on neovim nightly `v0.5.0-428-g46fdad255` fwiw.

I'd call this solved (: 

you the bomb lacambre
